source/common/conn_pool/conn_pool_base.cc

```diff
@@ -279,11 +285,15 @@ ConnectionPool::Cancellable* ConnPoolImplBase::newStreamImpl(AttachContext& cont
                  result == ConnectionResult::FailedToCreateConnection),
             fmt::format("Failed to create expected connection: {}", *this));
   if (result == ConnectionResult::FailedToCreateConnection) {
-    // This currently only happens for HTTP/3 if secrets aren't yet loaded.
-    // Trigger connection failure.
-    pending->cancel(Envoy::ConnectionPool::CancelPolicy::CloseExcess);
-    onPoolFailure(nullptr, absl::string_view(), ConnectionPool::PoolFailureReason::Overflow,
-                  context);
+    // This currently only happens for HTTP/3 if 1) secrets aren't yet loaded,
+    // or 2) connect failed immediately.
+    if (!pending_streams_.empty()) {
+      // If pending streams hasn't been purged. This failure must be because of 1). Trigger
+      // connection failure.
+      pending->cancel(Envoy::ConnectionPool::CancelPolicy::CloseExcess);
+      onPoolFailure(nullptr, absl::string_view(), ConnectionPool::PoolFailureReason::Overflow,
+                    context);
+    }
```

Comment:
Can we add unit tests for the new code in this file?

Reply:
Added a test in conn_pool_base_test.cc to verify the new way purges all pending streams.